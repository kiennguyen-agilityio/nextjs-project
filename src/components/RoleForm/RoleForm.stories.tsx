import type { Meta, StoryObj } from '@storybook/react';

// components
import RoleForm from '@/components/RoleForm';

const meta = {
  title: 'Components/RoleForm',
  component: RoleForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof RoleForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
