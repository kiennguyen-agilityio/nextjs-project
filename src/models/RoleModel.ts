export type RoleModel = {
  id: string;
  name: string;
  description: string;
  createAt?: Date;
};
