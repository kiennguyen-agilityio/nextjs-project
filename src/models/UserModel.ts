export interface UserModel {
  id: string;
  name: string;
  email: string;
  userRole: number;
  avatar: string;
  joined: Date;
}
