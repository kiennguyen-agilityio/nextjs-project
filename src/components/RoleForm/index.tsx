'use client';
import { useState } from 'react';

// models

// components
import { Input } from '@/components/common/Input';
import { Button } from '@/components/common/Button';

// models
import { RoleModel } from '@/models/RoleModel';

export interface RoleFormProps {
  role?: RoleModel;
}

const RoleForm = ({ role }: RoleFormProps) => {
  const [title] = useState(role?.name || 'Admin');
  const [description] = useState(
    role?.description || 'Admin is the super user',
  );

  const handleOnchange = () => {
    // TODO: handle onchange later
  };

  return (
    <form className="max-w-md mx-auto p-4 bg-white shadow-md rounded-lg">
      <h2 className="text-xl font-bold mb-4">Role</h2>
      <p className="text-sm text-gray-600 mb-4">
        Everyone who works on your {title} can have different roles
        <a href="#" className="text-blue-500 ml-1">
          learn more
        </a>
      </p>
      <div className="mb-4">
        <Input
          type="role"
          name="title"
          label="Type name"
          onChange={handleOnchange}
        />
      </div>

      <div className="mb-4">
        <label className="block text-sm font-medium text-gray-700">
          Description
        </label>
        <textarea className="mt-1 p-2 w-full border rounded-md" rows={4}>
          {description}
        </textarea>
      </div>
      <div className="flex justify-end space-x-2">
        <Button customClass="px-4 py-2 border rounded-md" variant="outline">
          Cancel
        </Button>
        <button className="px-4 py-2 bg-blue-500 text-white rounded-md">
          Create role
        </button>
      </div>
    </form>
  );
};

export default RoleForm;
