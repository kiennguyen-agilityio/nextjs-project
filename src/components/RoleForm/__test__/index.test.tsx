import { render, screen, fireEvent } from '@testing-library/react';
import RoleForm from '@/components/RoleForm';
import { RoleModel } from '@/models/RoleModel';

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  useFormState: () => [{}, () => {}, false],
  useFormStatus: () => ({ pending: false }),
}));

const role: RoleModel = {
  id: '1',
  name: 'Admin',
  description: 'Administrator role with full permissions',
};

// Reset mocks before each test
beforeEach(() => {
  jest.resetAllMocks();
});

describe('RoleForm', () => {
  test('renders empty role form', () => {
    const { asFragment } = render(<RoleForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders role form with initial data', () => {
    render(<RoleForm id={role.id} role={role} />);
    expect(screen.getByLabelText('Type name')).toHaveValue(role.name);
    expect(screen.getByLabelText('Description')).toHaveValue(role.description);
  });

  test('handles input change for name field', () => {
    render(<RoleForm id={role.id} role={role} />);
    const nameInput = screen.getByLabelText('Type name');
    fireEvent.change(nameInput, { target: { value: 'Manager' } });
    expect(nameInput).toHaveValue('Manager');
  });

  test('handles input change for description field', () => {
    render(<RoleForm id={role.id} role={role} />);
    const descriptionTextarea = screen.getByLabelText('Description');
    fireEvent.change(descriptionTextarea, {
      target: { value: 'Manager role' },
    });
    expect(descriptionTextarea).toHaveValue('Manager role');
  });

  test('submit button is disabled if form has not changed', () => {
    render(<RoleForm id={role.id} role={role} />);
    const submitButton = screen.getByText('Update Role');
    expect(submitButton).toBeDisabled();
  });

  test('submit button is enabled if form has changed', () => {
    render(<RoleForm id={role.id} role={role} />);
    const nameInput = screen.getByLabelText('Type name');
    fireEvent.change(nameInput, { target: { value: 'Manager' } });
    const submitButton = screen.getByText('Update Role');
    expect(submitButton).toBeEnabled();
  });
});
