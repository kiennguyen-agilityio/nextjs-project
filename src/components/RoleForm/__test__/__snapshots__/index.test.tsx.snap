// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoleForm renders empty role form 1`] = `
<DocumentFragment>
  <form
    class="max-w-full md:max-w-2xl lg:max-w-4xl xl:max-w-full mx-auto p-4 bg-white shadow-md rounded-lg"
  >
    <h2
      class="text-xl font-bold mb-4"
    >
      Role
    </h2>
    <p
      class="text-sm text-gray-600 mb-4"
    >
      Everyone who works on your  can have different roles
      <a
        class="text-blue-500 ml-1"
        href="#"
      >
        learn more
      </a>
    </p>
    <div
      class="mb-4"
    >
      <div>
        <label
          class="block text-dark-100 text-sm font-medium leading-5 mb-4"
          for="name"
        >
          Type name
        </label>
        <input
          class="rounded-[10px] w-full h-12 py-2 px-4 bg-white-300 text-dark-90 placeholder-dark-50 leading-tight border border-gray-300 focus:outline-none focus:shadow-outline focus:ring-2 focus:ring-slate-300 "
          id="name"
          name="name"
          type="role"
          value=""
        />
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="name-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-gray-700"
        for="description"
      >
        Description
      </label>
      <textarea
        class="mt-1 p-2 w-full border rounded-md"
        id="description"
        name="description"
        rows="4"
      />
      <div
        aria-atomic="true"
        aria-live="polite"
        id="description-error"
      />
    </div>
    <div
      class="flex flex-col sm:flex-row sm:justify-between items-center space-y-2 sm:space-y-0 sm:space-x-2"
    >
      <div
        class="w-full sm:w-auto"
      >
        <a
          href="/roles"
        >
          <button
            aria-label="button"
            class="flex items-center rounded p-2 border border-gray-200 dark:border-gray-700 hover:bg-gray-100 w-full sm:w-auto px-4 py-2 border border-blue-500 text-blue-500 rounded-md"
            type="button"
          >
            <span
              class="mr-2"
            >
              <svg
                class="w-4 h-4"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 15 3 9m0 0 6-6M3 9h12a6 6 0 0 1 0 12h-3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            Back to Role List
          </button>
        </a>
      </div>
      <div
        class="flex flex-col sm:flex-row sm:space-x-2 space-y-2 sm:space-y-0 w-full sm:w-auto"
      >
        <a
          href="/roles"
        >
          <button
            aria-label="button"
            class="flex items-center rounded p-2 border border-gray-200 dark:border-gray-700 hover:bg-gray-100 w-full sm:w-auto px-4 py-2 border rounded-md"
            type="button"
          >
            Cancel
          </button>
        </a>
        <button
          aria-label="button"
          class="flex items-center rounded p-2 bg-[#4270ec] text-white hover:cursor-not-allowed hover:opacity-70 px-4 py-2 bg-blue-500 text-white rounded-md"
          disabled=""
          type="submit"
        >
          Create Role
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
